File: sources/relational/debezium-reactivator/src/main/java/com/drasi/Reactivator.java
/*
 * Copyright 2024 The Drasi Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.drasi;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.sql.SQLException;
import java.util.HashMap;
import java.util.Map;

@SpringBootApplication
public class Reactivator {

    private static final Logger log = LoggerFactory.getLogger(Reactivator.class);

    public static void main(String[] args) throws IOException, SQLException {

        try {
            var sourceId = System.getenv("SOURCE_ID");
            var connector = System.getenv("connector");
            var instanceId = System.getenv("INSTANCE_ID");
            SpringApplication app = new SpringApplication(Reactivator.class);
            Map<String, Object> defaultProperties = new HashMap<>();
            defaultProperties.put("server.port", "80");
            defaultProperties.put("drasi.sourceid", sourceId);
            defaultProperties.put("drasi.connector", connector);
            defaultProperties.put("drasi.instanceid", instanceId);
            defaultProperties.put("server.shutdown", "graceful");
            defaultProperties.put("spring.lifecycle.timeout-per-shutdown-phase", "20s");
            app.setDefaultProperties(defaultProperties);
            app.run(args);
        } catch (Exception ex) {
            log.error(ex.getMessage());
            Files.write(Path.of("/dev/termination-log"), ex.getMessage().getBytes());
            throw ex;
        }
    }



}

File: sources/sdk/dotnet/Drasi.Source.SDK/Reactivator.cs
// Copyright 2024 The Drasi Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System.Diagnostics;
using Microsoft.AspNetCore.Builder;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;

namespace Drasi.Source.SDK;

public class Reactivator : IHost
{
    private static ActivitySource _traceSource = new("Drasi.Reactivator");
    private readonly WebApplication _app;

    public IServiceProvider Services => _app.Services;
    public IConfiguration Configuration => _app.Configuration;

    public ILogger<Reactivator> Logger => _app.Services.GetRequiredService<ILogger<Reactivator>>();

    internal Reactivator(WebApplication app)
    {
        _app = app;
        var deprovisionHandler = _app.Services.GetService<IDeprovisionHandler>();
        if (deprovisionHandler != null)
        {
            app.MapPost("deprovision", async context =>
            {
                await deprovisionHandler.Deprovision(_app.Services.GetRequiredService<IStateStore>());
                context.Response.StatusCode = 200;
            });
        }
    }

    public async Task StartAsync(CancellationToken cancellationToken = default)
    {
        Logger.LogInformation("Starting reactivator");

        var cts = new CancellationTokenSource();
        _ = _app.WaitForShutdownAsync(cancellationToken).ContinueWith(t => cts.Cancel(), cancellationToken);
        _ = _app.RunAsync($"http://0.0.0.0:{Configuration["APP_PORT"] ?? "80"}");
        var changeMonitor = _app.Services.GetRequiredService<IChangeMonitor>();
        var publisher = _app.Services.GetRequiredService<IChangePublisher>();

        try
        {
            await foreach (var change in changeMonitor.Monitor(cts.Token))
            {
                using var activity = _traceSource.StartActivity("PublishChange");
                var reactivator_end_time = (DateTimeOffset.UtcNow.Ticks - DateTimeOffset.UnixEpoch.Ticks) * 100;
                change.SetReactivatorEndNs(reactivator_end_time);
                await publisher.Publish(change);
            }
        }
        catch (OperationCanceledException)
        {
        }
    }

    public async Task StopAsync(CancellationToken cancellationToken = default)
    {
        Logger.LogInformation("Stopping reactivator");
        await _app.StopAsync(cancellationToken);
    }

    public void Dispose()
    {

    }


    public static void TerminateWithError(string message)
    {
        Console.WriteLine(message);
        try
        {
            File.WriteAllText("/dev/termination-log", message);
        }
        finally
        {
            Environment.Exit(1);
        }
    }

    public static string SourceId()
    {
        return Environment.GetEnvironmentVariable("SOURCE_ID");
    }
}